src/routes/slack.js

```diff
@@ -0,0 +1,183 @@
+/* @flow */
+
+const logger = require('../lib/logger');
+
+const rp = require('request-promise');
+const router = require('express').Router();
+
+const SlackCredentialModel = require('../models/SlackCredentialModel');
+const SlackCredential = require('../schemas/SlackCredential');
+
+const SongSearch = require('../models/SongSearch');
+const SongTemplate = require('../templates/slack/SongTemplate');
+
+const ArtistSearch = require('../models/ArtistSearch');
+const ArtistTemplate = require('../templates/slack/ArtistTemplate');
+
+const AlbumSearch = require('../models/AlbumSearch');
+const AlbumTemplate = require('../templates/slack/AlbumTemplate');
+
+const CardTemplate = require('../templates/slack/CardTemplate');
+const ResponseTemplate = require('../templates/slack/ResponseTemplate');
+
+const validParams = ['song', 'artist', 'album'];
+
+/**
+ * Middleware layer that verifies that both parameter and keyword
+ * are present in the command
+ * @type {void}
+ */
+router.post('/incoming', (req, res, next) => {
```

Comment:
Let me propose something here:
```js
const getTokens = text => text.split(' ');
const getSearchParameter = text => getTokens(text)[0];
const isValidSearchParameter = text => validParams.includes(getSearchParameter(text));
const isValidKeyword = text => buildKeyword(text);
const buildKeyword = text => getTokens(text).slice(1).join(' ');

const decorateRequest = ({text, user_name: userName}) => {
  // I proposed this new key instead of user, because user could sound like there's an instance of type user
  req.userName = userName;
  // also proposed searchParam instead of param to make it explicit and more understandable
  req.searchParam = getSearchParameter(text);
  req.keyword = buildKeyword(text);
};

router.post('/incoming', (req, res, next) => {
  const { body } = req;
  const { text, user_name: userName } = body;

  if (!isValidSearchParameter(text) || !isValidKeyword(text)) {
    return res.status(200).send('Please enter a valid command!');
  } 

  decorateRequest(body);
});
```

This proposal abstracts the low level details into functions, so whenever I come and read the middleware, I just read the algorithm without the need of knowing too much details of tokens and stuff. I just read:
if the request is sending me an invalid search parameter or the keyword is also invalid, I say "fuck off" to the user, if not, I just decorate the request.
PD: Low level details are stuff like `text.split(' ').splice(1)[3]`, something that is very hard to read and is not relevant for the reader.